channeldb/control_tower.go

```diff
@@ -0,0 +1,424 @@
+// ControlTower interface provides access to driving the state transitions.
+type ControlTower interface {
+	// InitPayment atomically moves the payment into the InFlight state.
+	// This method checks that no completed payment exist for this payment
+	// hash.
+	InitPayment(lntypes.Hash, *PaymentCreationInfo) error
+
+	// RegisterAttempt atomically records the provided PaymentAttemptInfo.
+	RegisterAttempt(lntypes.Hash, *PaymentAttemptInfo) error
+
+	// Success transitions a payment into the Completed state. After
+	// invoking this method, InitPayment should always return an error to
+	// prevent us from making duplicate payments to the same payment hash.
+	// The provided preimage is atomically saved to the DB for record
+	// keeping.
+	Success(lntypes.Hash, lntypes.Preimage) error
+
+	// Fail transitions a payment into the Failed state, and records the
+	// reason the payment failed. After invoking this method, InitPayment
+	// should return nil on its next call for this payment hash, allowing
+	// the switch to make a subsequent payment.
+	Fail(lntypes.Hash, FailureReason) error
+
+	// FetchInFlightPayments returns all payments with status InFlight.
+	FetchInFlightPayments() ([]*InFlightPayment, error)
+}
+
+// paymentControl is persistent implementation of ControlTower to restrict
+// double payment sending.
+type paymentControl struct {
+	db *DB
+}
+
+// NewPaymentControl creates a new instance of the paymentControl.
+func NewPaymentControl(db *DB) ControlTower {
+	return &paymentControl{
+		db: db,
+	}
+}
+
+// InitPayment checks or records the given PaymentCreationInfo with the DB,
+// making sure it does not already exist as an in-flight payment. Then this
+// method returns successfully, the payment is guranteeed to be in the InFlight
+// state.
+func (p *paymentControl) InitPayment(paymentHash lntypes.Hash,
+	info *PaymentCreationInfo) error {
+
+	var b bytes.Buffer
+	if err := serializePaymentCreationInfo(&b, info); err != nil {
+		return err
+	}
+	infoBytes := b.Bytes()
+
+	var takeoffErr error
+	err := p.db.Batch(func(tx *bbolt.Tx) error {
+		bucket, err := fetchPaymentBucket(tx, paymentHash)
+		if err != nil {
+			return err
+		}
+
+		// Get the existing status of this payment, if any.
+		paymentStatus := fetchPaymentStatus(bucket)
+
+		// Reset the takeoff error, to avoid carrying over an error
+		// from a previous execution of the batched db transaction.
+		takeoffErr = nil
+
+		switch paymentStatus {
+
+		// We allow retrying failed payments.
+		case StatusFailed:
+
+		// This is a new payment that is being initialized for the
+		// first time.
+		case StatusGrounded:
+
+		// We already have an InFlight payment on the network. We will
+		// disallow any new payments.
+		case StatusInFlight:
+			takeoffErr = ErrPaymentInFlight
+			return nil
+
+		// We've already completed a payment to this payment hash,
+		// forbid the switch from sending another.
+		case StatusCompleted:
```

Comment:
Reminder: rename to StatusSucceeded to avoid ambiguity. "Completed" is also used in other places in this pr.

Reply:
Done. Also renamed `Grounded->Unknown`: https://github.com/lightningnetwork/lnd/pull/2761/commits/29fce7d61820d2fb85209f1fd4583bc78a7a61f7